### PR TITLE
Extend input API utilities to support input operators

### PR DESCRIPTION
**Pull request by szalpal**
Signed-off-by: szalpal <mszolucha@nvidia.com>

## Category:

**New feature** (*non-breaking change which adds functionality*)

## Description:

This PR is a next one in the effort of introducing input operators. We had some API utilities that supported only external source. It was hardcoded. Now, since `external_source` is also an input operator, these utilities can and should support all input operators.

There is a minor change in the implementation. Since we shouldn't explore the operators by name now, I assumed that these utilities may be called provided the graph has been built. Therefore I can use dynamic casting to the InputOperator base class to determine, whether an operator in an input operator or not.

* Make the functions: layout(), ndim(), dtype(), num_inputs() support all input operators.

### Tests:

- [ ] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build

File: dali/pipeline/pipeline.cc [hunk 1]
```diff
@@ -873,10 +895,24 @@ int Pipeline::output_ndim(int id) const {
...
+         (op_node.op_type == OpType::GPU &&
+          dynamic_cast<InputOperator<GPUBackend> *>(op_node.op.get()));
+}
+
+
 int Pipeline::num_inputs() const {
-  return ext_input_names_.size();
+  DALI_ENFORCE(built_, "\"Build()\" must be called prior to calling \"num_inputs()\".");
+  auto &op_nodes = graph_.GetOpNodes();
+  return static_cast<int>(std::count_if(op_nodes.begin(), op_nodes.end(), is_input_operator));
```
Comment (jantonguirao) [inline dali/pipeline/pipeline.cc:912]: can we calculate this once and keep the result as a member? It could be calculated on pipeline build or similar.

Comment (jantonguirao) [reply dali/pipeline/pipeline.cc:912]: Also could you use:
`input_operators_names_.size()`?

**Review by jantonguirao — APPROVED** [review]

File: dali/pipeline/pipeline.cc (hunk 1 above)
Comment (szalpal) [reply dali/pipeline/pipeline.cc:912]: Yeah, seems like a good idea

**Review by stiepan — APPROVED** [review]

**Comment by szalpal** [comment]
!build

**Pull request merged**